--- DirectFB-1.6.2.org/lib/direct/os/linux/glibc/system.c
+++ DirectFB-1.6.2/lib/direct/os/linux/glibc/system.c
@@ -33,6 +33,7 @@
 #include <errno.h>
 #include <signal.h>
 #include <unistd.h>
+#include <sys/syscall.h>
 
 #include <linux/unistd.h>
 
@@ -107,7 +108,8 @@
 void
 direct_trap( const char *domain, int sig )
 {
-     sigval_t val;
+
+     union sigval val;
 
      D_LOG( Direct_Trap, VERBOSE, "Raising signal %d from %s...\n", sig, domain );
 
