--- screen-4.2.1.orig/configure	2014-04-26 12:58:46.000000000 +0200
+++ screen-4.2.1/configure	2014-05-01 13:04:48.000000000 +0200
@@ -414,7 +414,7 @@ as_fn_error ()
     $as_echo "$as_me:${as_lineno-$LINENO}: error: $2" >&$4
   fi
   $as_echo "$as_me: error: $2" >&2
-  as_fn_exit $as_status
+  # as_fn_exit $as_status
 } # as_fn_error
 
 if expr a : '\(a\)' >/dev/null 2>&1 &&
@@ -5941,8 +5941,6 @@ else
 fi
 rm -f core conftest.err conftest.$ac_objext \
     conftest$ac_exeext conftest.$ac_ext
-test -f /lib/libsec.a || test -f /usr/lib/libsec.a && LIBS="$LIBS -lsec"
-test -f /lib/libshadow.a || test -f /usr/lib/libshadow.a && LIBS="$LIBS -lshadow"
 oldlibs="$LIBS"
 LIBS="$LIBS -lsun"
 { $as_echo "$as_me:${as_lineno-$LINENO}: checking IRIX sun library..." >&5
@@ -7023,7 +7021,7 @@ as_fn_error ()
     $as_echo "$as_me:${as_lineno-$LINENO}: error: $2" >&$4
   fi
   $as_echo "$as_me: error: $2" >&2
-  as_fn_exit $as_status
+  # as_fn_exit $as_status
 } # as_fn_error
 
 
