--- node-v0.10.2.org/deps/v8/src/atomicops.h
+++ node-v0.10.2/deps/v8/src/atomicops.h
@@ -69,7 +69,7 @@
 
 // Use AtomicWord for a machine-sized pointer.  It will use the Atomic32 or
 // Atomic64 routines below, depending on your architecture.
-#if defined(__OpenBSD__) && defined(__i386__)
+#if defined(__i386__)
 typedef Atomic32 AtomicWord;
 #else
 typedef intptr_t AtomicWord;
--- node-v0.10.2.org/deps/v8/src/platform.h
+++ node-v0.10.2/deps/v8/src/platform.h
@@ -107,7 +107,7 @@
 
 // Use AtomicWord for a machine-sized pointer. It is assumed that
 // reads and writes of naturally aligned values of this type are atomic.
-#if defined(__OpenBSD__) && defined(__i386__)
+#if defined(__i386__)
 typedef Atomic32 AtomicWord;
 #else
 typedef intptr_t AtomicWord;
