--- mozilla-release.org/nsprpub/pr/src/misc/prnetdb.c	2013-10-25 22:27:38.000000000 +0000
+++ mozilla-release/nsprpub/pr/src/misc/prnetdb.c	2013-10-31 00:53:23.677525286 +0000
@@ -89,6 +89,11 @@
 #define _PR_HAVE_5_ARG_GETPROTO_R
 #endif
 
+#undef _PR_HAVE_GETPROTO_R
+#undef _PR_HAVE_GETPROTO_R_INT
+#undef _PR_HAVE_GETPROTO_R_POINTER
+#undef _PR_HAVE_5_ARG_GETPROTO_R
+
 #if !defined(_PR_HAVE_GETPROTO_R)
 PRLock* _getproto_lock = NULL;
 #endif
