--- mozilla-release.org/configure
+++ mozilla-release/configure
@@ -25848,7 +25848,7 @@
   fi
 fi
 
-NONASCII=$'\241\241'
+NONASCII='\241\241'
 
 
 # Static build options
@@ -28922,7 +28922,7 @@
     (''' EXPAND_LIBS_LIST_STYLE ''', r''' $EXPAND_LIBS_LIST_STYLE ''')
     (''' EXPAND_LIBS_ORDER_STYLE ''', r''' $EXPAND_LIBS_ORDER_STYLE ''')
     (''' CL_INCLUDES_PREFIX ''', r''' $CL_INCLUDES_PREFIX ''')
-    (''' NONASCII ''', r''' $NONASCII ''')
+    (''' NONASCII ''', r''' "$NONASCII" ''')
     (''' LIBXUL_LIBS ''', r''' $LIBXUL_LIBS ''')
     (''' GKMEDIAS_SHARED_LIBRARY ''', r''' $GKMEDIAS_SHARED_LIBRARY ''')
     (''' ZLIB_IN_MOZGLUE ''', r''' $ZLIB_IN_MOZGLUE ''')
--- mozilla-release.org/js/src/configure
+++ mozilla-release/js/src/configure
@@ -14858,7 +14858,7 @@
   fi
 fi
 
-NONASCII=$'\241\241'
+NONASCII='\241\241'
 
 
 if test -n "$GNU_CC" -a -z "$CLANG_CC"; then
@@ -16451,7 +16451,7 @@
     (''' EXPAND_LIBS_LIST_STYLE ''', r''' $EXPAND_LIBS_LIST_STYLE ''')
     (''' EXPAND_LIBS_ORDER_STYLE ''', r''' $EXPAND_LIBS_ORDER_STYLE ''')
     (''' CL_INCLUDES_PREFIX ''', r''' $CL_INCLUDES_PREFIX ''')
-    (''' NONASCII ''', r''' $NONASCII ''')
+    (''' NONASCII ''', r''' "$NONASCII" ''')
     (''' JS_NATIVE_EDITLINE ''', r''' $JS_NATIVE_EDITLINE ''')
     (''' JS_DISABLE_SHELL ''', r''' $JS_DISABLE_SHELL ''')
     (''' EDITLINE_LIBS ''', r''' $EDITLINE_LIBS ''')
